sdk/storage/Azure.Storage.Files.Shares/perf/Azure.Storage.Files.Shares.Perf/Scenarios/ShareClientManager.cs

```diff
@@ -0,0 +1,92 @@
+﻿//Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Test.Perf;
+
+namespace Azure.Storage.Files.Shares.Perf.Scenarios
+{
+    internal sealed class ShareClientManager : IAsyncDisposable
```

Comment:
This reminds me about [this](https://www.petrikainulainen.net/programming/unit-testing/3-reasons-why-we-should-not-use-inheritance-in-our-tests/). I've seen couple of test frameworks turning into a hairball (all used inheritance to share code).

Reply:
I see both sides, personally I think inheritance works well in this specific case, but I'm fine with whatever the language team prefers.  I would like to be consistent across tests but we can also just leave this up to the preference of whomever writes the tests.

@pakrym: Do we have any guidance about code reuse and common base classes in our unit tests?  Whatever guidance we use here should apply to the perf tests as well.

Reply:
I agree inheritance makes the tests harder to read (you need to jump through the base classes).  But it also encapsulates the code into logical layers, and it avoids needing to fix the same test bug in multiple places.  To me, it makes more sense to have a base class responsible for creating/deleting the share, rather than duplicate this code in every test.

Reply:
The DRY can be achieved by [composition](https://en.wikipedia.org/wiki/Composition_over_inheritance) as well. Which can also get messy but imo it's easier to refactor if that happens.
Inheritance is layers, composition is lego bricks. I'd rather play with lego.

Reply:
I agree inheritance and composition each have their place.  This is getting into the realm of style/preference, so I think the perf tests should just align with whatever guidelines we do or don't have for similar concerns in the other code in the repo.  If there is no guidance then I think whomever "owns" the code for a particular SDK can decide.

Reply:
>@pakrym: Do we have any guidance about code reuse and common base classes in our unit tests? Whatever guidance we use here should apply to the perf tests as well.

There is no overall guidance. The Azure.Core.TestFramework promotes inheritance and most of the test projects I've seen use it as well.

Reply:
To not get into extreme. sometimes right mix of the two is ok-ish. I.e. hooking up some behaviors via inheritance but delegating major portion of implementation to collaborating class. So that base types don't get bloated over time.

For example we have DisposingFileSystem (see [here](https://github.com/Azure/azure-sdk-for-net/blob/1041c993391cb4b2dfee3d880514f2b0d95e1c41/sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs#L1128) ). And some methods in base class to obtain it. (Not saying that lifecycle here is perfect)

I think `ShareClientManager` could become `DisposableTestShare` and base test class can spin this up and then dispose. So that lifecycle of that thing is handled by base class, but implementation goes outside of base class. Test writer would then access a property with `DisposableTestShare` in the scenario.
We do something similar with `TestEnvironment` don't we?

Reply:
@kasobol-msft, @Mohit-Chakraborty: I think how to share this common code is up to you two.  I will remove myself from the decision :smile:.

Reply:
Right. I definitely agree there is a balance to it and not everything has to go into the base class. The test framework does provide multiple extension points that allow plugging in implementation for sanitizers, request matchers, test environment, etc.

To me, the question is whether the base implementation is useful for most of the classes that inherit from it so it doesn't become a dumping ground for disconnected utility methods.